x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ForceMergeStep.java

```diff
@@ -39,10 +48,29 @@ public int getMaxNumSegments() {
     @Override
     public void performAction(IndexMetadata indexMetadata, ClusterState currentState,
                               ClusterStateObserver observer, ActionListener<Boolean> listener) {
-        ForceMergeRequest request = new ForceMergeRequest(indexMetadata.getIndex().getName());
+        String indexName = indexMetadata.getIndex().getName();
+        ForceMergeRequest request = new ForceMergeRequest(indexName);
         request.maxNumSegments(maxNumSegments);
         getClient().admin().indices()
-            .forceMerge(request, ActionListener.wrap(response -> listener.onResponse(true),
+            .forceMerge(request, ActionListener.wrap(
+                response -> {
+                    if (response.getFailedShards() == 0) {
+                        listener.onResponse(true);
+                    } else {
+                        DefaultShardOperationFailedException[] failures = response.getShardFailures();
+                        String policyName = LifecycleSettings.LIFECYCLE_NAME_SETTING.get(indexMetadata.getSettings());
+                        String errorMessage =
+                            String.format(Locale.ROOT, "index [%s] in policy [%s] encountered failures [%s] on step [%s]",
+                                indexName, policyName,
+                                failures == null ? "n/a" : Strings.collectionToDelimitedString(Arrays.stream(failures)
+                                    .map(Strings::toString)
+                                    .collect(Collectors.toList()), ","),
+                                NAME);
+                        logger.warn(errorMessage);
+                        // let's report it as a failure and retry
+                        listener.onFailure(new IllegalStateException(errorMessage));
```

Comment:
I'm trying to think if there is a better exception type to use here.. maybe we should use `ElasticsearchException` with your current message and the first `DefaultShardOperationFailedException` passed as the `cause`.

What do you think?

Reply:
I changed it to use `ElasticsearchException` but very misleadingly `DefaultShardOperationFailedException` is **not** an `Exception` so we'll have to carry on with a null `cause`

Reply:
Okay, that's unfortunate naming indeed, but thanks for the explanation :)